### PR TITLE
feat(server): add delete endpoint

### LINKED ISSUE #127: DELETE end point
I'm using rustypaste on my LAN as a lightweight file server for transferring files/links/text across devices (currently I'm building the UI for it). One thing I realized rustypaste lacks which I would like to have is the ability to remove files and the ability to update files (change their contents or set a new expiration).

However, I'm aware that my use case may be niche and probably not something that would be included in most pastebin services, so before I set about building a PR for this I thought I'd see if this fits with the vision of the app enough to consider. If not I may just need to find a new backend service for my UI.

Interestingly I have found that there's not a lot out there in the ways of a simple file hosting service that have simple API access without a lot of other bells and whistles that I don't need. Most solutions are pretty heavy (like syncthing) which is why I ended up looking at pastebin services. Rustypaste fits quite nice in this regard with it's simple HTTP interface and speed. 

EDIT: thinking more on this, I think the only one I would truly need is DELETE.

### PR DESCRIPTION
**Pull request by ajbdev**
## Description

Add a DELETE endpoint guarded by a separate auth token array.

## Motivation and Context

Closes #127 

## How Has This Been Tested?

Wrote a test case and tested manually.

## Changelog Entry

````
### Added

- Change `Config.get_tokens` to `Config.get_auth_tokens` and update invocations.
- Add delete_tokens array to config and `get_delete_tokens` method to Config for retrieving delete tokens.
- Add DELETE endpoint that will remove files if a delete token is set and authorized by the request.
- Test for new DELETE endpoint
- Add DELETE_TOKEN_ENV for setting the delete tokens array from an ENV var.
````
-->

## Types of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (no code change)
- [ ] Refactor (refactoring production code)
- [ ] Other 

## Checklist:

- [x] My code follows the code style of this project.
- [ ] I have updated the documentation accordingly.
- [x] I have formatted the code with [rustfmt](https://github.com/rust-lang/rustfmt).
- [x] I checked the lints with [clippy](https://github.com/rust-lang/rust-clippy).
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by tessus — COMMENTED** [review]
LGTM

closes #127

**Comment by tessus** [comment]
I was thinking a bit more. Maybe we should create an enum for token_type: auth, delete. And then use just one function `get_tokens(&self, enum)`. I have no idea, whether this is better code or me just trying to optimize without any gain.

For me thia PR is perfecctly fine. Let's see what @orhun has to say.

File: README.md [hunk 1]
```diff
@@ -79,6 +80,7 @@ Here you can read the blog post about how it is deployed on Shuttle: [https://bl
     - supports forcing to download via `?download=true`
   - no duplicate uploads (optional)
   - listing files
+  - deleting files
```
Comment (orhun) [inline README.md:83]:
```suggestion
  - listing/deleting files
```

File: README.md [hunk 2]
```diff
@@ -235,6 +237,14 @@ while read -r filename; do
 done
 ```
 
+#### Delete file from server
+
+Set `delete_tokens` array in [config.toml](./config.toml) to activate the DELETE endpoint and secure it with one (or more) auth token(s).
```
Comment (orhun) [inline README.md:242]:
```suggestion
Set `delete_tokens` array in [config.toml](./config.toml) to activate the [`DELETE`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/DELETE) endpoint and secure it with one (or more) auth token(s).
```

File: src/server.rs [hunk 3]
```diff
@@ -146,6 +146,33 @@ async fn serve(
...
+    let tokens = config.get_delete_tokens();
+    if tokens.is_none() {
+        log::warn!("delete endpoint not served because there are no delete_tokens set");
+        return Err(error::ErrorForbidden("endpoint is not exposed\n"));
+    }
+    auth::check(host, request.headers(), tokens)?;
+    if !path.is_file() || !path.exists() {
+        return Err(error::ErrorNotFound("file is not found or expired :(\n"));
+    }
+    fs::remove_file(path).expect("Could not delete file");
```
Comment (orhun) [inline src/server.rs:172]: Can we use `error!` here to log the error and return 500?

File: src/server.rs [hunk 4]
```diff
@@ -146,6 +146,33 @@ async fn serve(
...
+    let config = config
+        .read()
+        .map_err(|_| error::ErrorInternalServerError("cannot acquire config"))?;
+    let path = config.server.upload_path.join(&*file);
+    let path = util::glob_match_file(path)?;
+    let connection = request.connection_info().clone();
+    let host = connection.realip_remote_addr().unwrap_or("unknown host");
+    let tokens = config.get_delete_tokens();
+    if tokens.is_none() {
+        log::warn!("delete endpoint not served because there are no delete_tokens set");
```
Comment (orhun) [inline src/server.rs:165]: Are we testing this case with unit tests/fixtures? (not strictly needed though)

**Review by orhun — COMMENTED** [review]
Overall looks good to me. Just had a few comments 🐻 

@tessus I don't know what would be the benefit of adding an `enum` but we can go with that for the sake of having a single function for retrieving the tokens. @ajbdev can you also make that change?

**Comment by tessus** [comment]
> @tessus I don't know what would be the benefit of adding an enum but we can go with that for the sake of having a single function for retrieving the tokens. @ajbdev can you also make that change?

I came up with this "brilliant" idea, thus I'll take care of it. ;-)

File: src/server.rs (hunk 3 above)
Comment (tessus) [reply src/server.rs]: done

File: src/server.rs (hunk 4 above)
Comment (tessus) [reply src/server.rs:165]: done

**Comment by tessus** [comment]
@orhun all done!

**Comment by tessus** [comment]
@orhun It seems that gh had a connection issue during 2 tests. Any chance you can rerun the action?

If I run the tests locally, everything is fine.

**Review by orhun — APPROVED** [review]
Tested and works well!

Thank you both @ajbdev @tessus 🚀

**Pull request merged**